### PR TITLE
CAT-1360 SetColor and ChangeColor Bricks

### PR DESCRIPTION
**Pull request by Gjallar**
#CAT-1360
https://jira.catrob.at/browse/CAT-1360

### CODE REVIEW

File: catroid/src/org/catrobat/catroid/content/actions/ChangeColorByNAction.java [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+import org.catrobat.catroid.formulaeditor.InterpretationException;
+
+public class ChangeColorByNAction extends TemporalAction {
+
+	private Sprite sprite;
+	private Formula color;
+
+	protected void update(float delta) {
+		try {
+			Float eightBitColor = color == null ? 25 : color.interpretFloat(sprite);
```
Comment (thmq) [inline catroid/src/org/catrobat/catroid/content/actions/ChangeColorByNAction.java]: why you no float?

File: catroid/src/org/catrobat/catroid/content/Look.java [hunk 2]
```diff
@@ -370,6 +379,24 @@ public void changeBrightnessInUserInterfaceDimensionUnit(float changePercent) {
 		setBrightnessInUserInterfaceDimensionUnit(getBrightnessInUserInterfaceDimensionUnit() + changePercent);
 	}
 
+	public float getColorInUserInterfaceDimensionUnit() {
+		return hue * 200.0f;
```
Comment (fwinkelbauer) [inline catroid/src/org/catrobat/catroid/content/Look.java]: 200 shows at several places in two methods:

- `getColorInUserInterfaceDimensionUnit`
- `setColorInUserInterfaceDimensionUnit`

Could you create a constant for that please?

**Comment by fwinkelbauer** [comment]
Could you please run the Android Studio Formatter on your changes? (See first bullet point [here](https://github.com/Catrobat/Catroid/wiki/Creating-a-pull-request)

File: catroid/src/org/catrobat/catroid/content/actions/ChangeColorByNAction.java [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+	private Sprite sprite;
+	private Formula color;
+
+	protected void update(float delta) {
+		try {
+			float eightBitColor = color == null ? 25 : color.interpretFloat(sprite);
+			sprite.look.changeColorInUserInterfaceDimensionUnit(eightBitColor);
+		} catch (InterpretationException interpretationException) {
+			Log.d(getClass().getSimpleName(), "Formula interpretation for this specific Brick failed.", interpretationException);
```
Comment (thomasmauerhofer) [inline catroid/src/org/catrobat/catroid/content/actions/ChangeColorByNAction.java]: You can use Log.e(...) to send errormessages to the logger

File: catroid/src/org/catrobat/catroid/content/actions/SetColorToAction.java [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+
+	private Sprite sprite;
+	private Formula color;
+
+	protected void update(float delta) {
+		try {
+			float eightBitColor = color == null ? 0 : color.interpretFloat(sprite);
+			sprite.look.setColorInUserInterfaceDimensionUnit(eightBitColor);
+		} catch (InterpretationException interpretationException) {
+			Log.d(getClass().getSimpleName(), "Formula interpretation for this specific Brick failed.", interpretationException);
```
Comment (thomasmauerhofer) [inline catroid/src/org/catrobat/catroid/content/actions/SetColorToAction.java]: Log.e(...) can be used

File: catroid/src/org/catrobat/catroid/content/bricks/SetColorBrick.java [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+	public View getPrototypeView(Context context) {
+		prototypeView = View.inflate(context, R.layout.brick_set_color_to, null);
+		TextView textSetSizeTo = (TextView) prototypeView.findViewById(R.id.brick_set_color_prototype_text_view);
+		textSetSizeTo.setText(String.valueOf(BrickValues.SET_COLOR_TO));
+		return prototypeView;
+	}
+
+	@Override
+	public View getViewWithAlpha(int alphaValue) {
+
```
Comment (thomasmauerhofer) [inline catroid/src/org/catrobat/catroid/content/bricks/SetColorBrick.java]: remove newline

File: catroid/src/org/catrobat/catroid/content/bricks/SetColorBrick.java [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+		if (view != null) {
+
```
Comment (thomasmauerhofer) [inline catroid/src/org/catrobat/catroid/content/bricks/SetColorBrick.java]: remove newline

File: catroid/src/org/catrobat/catroid/content/bricks/SetColorBrick.java [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+			TextView textSize = (TextView) view.findViewById(R.id.brick_set_color_label);
+			TextView editSize = (TextView) view.findViewById(R.id.brick_set_color_edit_text);
+			textSize.setTextColor(textSize.getTextColors().withAlpha(alphaValue));
+			editSize.setTextColor(editSize.getTextColors().withAlpha(alphaValue));
+			editSize.getBackground().setAlpha(alphaValue);
+
+			this.alphaValue = alphaValue;
+		}
+
```
Comment (thomasmauerhofer) [inline catroid/src/org/catrobat/catroid/content/bricks/SetColorBrick.java]: remove newline

File: catroid/src/org/catrobat/catroid/ui/fragment/FormulaEditorListFragment.java [hunk 8]
```diff
@@ -60,8 +60,8 @@
 
 	private static final int[] OBJECT_ITEMS = {R.string.formula_editor_object_x, R.string.formula_editor_object_y,
 			R.string.formula_editor_object_transparency, R.string.formula_editor_object_brightness,
-			R.string.formula_editor_object_size, R.string.formula_editor_object_rotation,
-			R.string.formula_editor_object_layer};
+			R.string.formula_editor_object_color, R.string.formula_editor_object_size, R.string.formula_editor_object_rotation,
+			R.string.formula_editor_object_layer };
```
Comment (thomasmauerhofer) [inline catroid/src/org/catrobat/catroid/ui/fragment/FormulaEditorListFragment.java]: remove space before }

File: catroidTest/src/org/catrobat/catroid/test/content/actions/ChangeColorByNActionTest.java [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+
+	private static final float INITIALIZED_VALUE = 0;
+	private static final String NOT_NUMERICAL_STRING = "color";
+	private static final float DELTA = 1;
+	private Sprite sprite;
+
+	@Override
+	protected void setUp() throws Exception {
+		sprite = new Sprite("testSprite");
+		super.setUp();
```
Comment (thomasmauerhofer) [inline catroidTest/src/org/catrobat/catroid/test/content/actions/ChangeColorByNActionTest.java]: please call super first

Comment (thomasmauerhofer) [inline catroidTest/src/org/catrobat/catroid/test/content/actions/ChangeColorByNActionTest.java]: please call super first

File: catroidTest/src/org/catrobat/catroid/test/content/actions/SetColorToTest.java [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+
+	private static final float COLOR = 100.0f;
+	private Formula color = new Formula(COLOR);
+	private static final String NOT_NUMERICAL_STRING = "NOT_NUMERICAL_STRING";
+	private Sprite sprite;
+
+	@Override
+	protected void setUp() throws Exception {
+		sprite = new Sprite("testSprite");
+		super.setUp();
```
Comment (thomasmauerhofer) [inline catroidTest/src/org/catrobat/catroid/test/content/actions/SetColorToTest.java]: please call super first

File: catroidTest/src/org/catrobat/catroid/uitest/content/brick/BrickValueParameterTest.java [hunk 11]
```diff
@@ -252,6 +252,42 @@ public void testLookBricksDefaultValues() {
 
 		assertEquals("Value in Selected Brick ChangeBrightness is not correct",
 				(float) BrickValues.CHANGE_BRITHNESS_BY, changeBrightnessEditTextValue);
+
+		UiTestUtils.clickOnBottomBar(solo, R.id.button_add);
+		solo.clickOnText(categoryLooksText);
+
+		// Just to get focus
```
Comment (thomasmauerhofer) [inline catroidTest/src/org/catrobat/catroid/uitest/content/brick/BrickValueParameterTest.java]: remove comment

File: catroidTest/src/org/catrobat/catroid/uitest/content/brick/ChangeColorByNBrickTest.java [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+		super(ScriptActivity.class);
+	}
+
+	@Override
+	public void setUp() throws Exception {
+		// normally super.setUp should be called first
+		// but kept the test failing due to view is null
+		// when starting in ScriptActivity
+		createProject();
+		super.setUp();
```
Comment (thomasmauerhofer) [inline catroidTest/src/org/catrobat/catroid/uitest/content/brick/ChangeColorByNBrickTest.java]: call super first

File: catroidTest/src/org/catrobat/catroid/uitest/content/brick/ChangeColorByNBrickTest.java [hunk 13]
```diff
@@ -0,0 +1,96 @@
...
+
+	public ChangeColorByNBrickTest() {
+		super(ScriptActivity.class);
+	}
+
+	@Override
+	public void setUp() throws Exception {
+		// normally super.setUp should be called first
+		// but kept the test failing due to view is null
+		// when starting in ScriptActivity
```
Comment (thomasmauerhofer) [inline catroidTest/src/org/catrobat/catroid/uitest/content/brick/ChangeColorByNBrickTest.java]: Can this be done by another way? Normally super.setUp() should always be called first, but there are also other tests which call it at end of method.... 
Remove the comment

File: catroidTest/src/org/catrobat/catroid/uitest/content/brick/ChangeColorByNBrickTest.java [hunk 14]
```diff
@@ -0,0 +1,96 @@
...
+		assertEquals("Incorrect number of bricks.", 1, childrenCount);
+
+		ArrayList<Brick> projectBrickList = project.getSpriteList().get(0).getScript(0).getBrickList();
+		assertEquals("Incorrect number of bricks.", 1, projectBrickList.size());
+
+		assertEquals("Wrong Brick instance.", projectBrickList.get(0), adapter.getChild(groupCount - 1, 0));
+		assertNotNull("TextView does not exist", solo.getText(solo.getString(R.string.brick_change_color)));
+
+		int newColor = 3;
+
```
Comment (thomasmauerhofer) [inline catroidTest/src/org/catrobat/catroid/uitest/content/brick/ChangeColorByNBrickTest.java]: remove newline

File: catroidTest/src/org/catrobat/catroid/uitest/content/brick/SetColorBrickTest.java [hunk 15]
```diff
@@ -0,0 +1,96 @@
...
+	private SetColorBrick setColorBrick;
+
+	public SetColorBrickTest() {
+		super(ScriptActivity.class);
+	}
+
+	@Override
+	public void setUp() throws Exception {
+		// normally super.setUp should be called first
+		// but kept the test failing due to view is null
```
Comment (thomasmauerhofer) [inline catroidTest/src/org/catrobat/catroid/uitest/content/brick/SetColorBrickTest.java]: same as above

File: catroidTest/src/org/catrobat/catroid/uitest/content/brick/SetColorBrickTest.java [hunk 16]
```diff
@@ -0,0 +1,96 @@
...
+		assertEquals("Incorrect number of bricks.", 1, childrenCount);
+
+		ArrayList<Brick> projectBrickList = project.getSpriteList().get(0).getScript(0).getBrickList();
+		assertEquals("Incorrect number of bricks.", 1, projectBrickList.size());
+
+		assertEquals("Wrong Brick instance.", projectBrickList.get(0), adapter.getChild(groupCount - 1, 0));
+		assertNotNull("TextView does not exist", solo.getText(solo.getString(R.string.brick_set_color)));
+
+		int newColor = 3;
+
```
Comment (thomasmauerhofer) [inline catroidTest/src/org/catrobat/catroid/uitest/content/brick/SetColorBrickTest.java]: remove newline

File: catroid/src/org/catrobat/catroid/content/actions/ChangeColorByNAction.java (hunk 3 above)
Comment (robertpainsi) [inline catroid/src/org/catrobat/catroid/content/actions/ChangeColorByNAction.java]: This line was copied from one of the many other actions. :wink: If the color formula doesn't evaluate to a number, the action is ignored. It's the expected behavior for this action and others like SetX, GoNStepsBackAction and so on.

So `Log.d` is fine. `Log.e` should only be used when something _fatal_ has happened and _won't be recoverable_.

https://source.android.com/source/code-style.html#log-sparingly, not our logging guidelines but still useful.
You may argue that it should not be logged at all, be logged as verbose instead or logging a stack trace is unnecessary. However, for consistency, I suggest to keep it the way it is. Feel free to create a Jira ticket which addresses _every_ action if you want to change something.

File: catroidTest/src/org/catrobat/catroid/uitest/content/brick/ChangeColorByNBrickTest.java (hunk 13 above)
Comment (robertpainsi) [inline catroidTest/src/org/catrobat/catroid/uitest/content/brick/ChangeColorByNBrickTest.java]: > Remove the comment

Personally, I like these type of comments. It explains why super.setUp() can't be called first. If this information is spread through the whole project, people don't have to face the same problem over and over again and it saves a lot of time.

Comment (fwinkelbauer) [inline catroidTest/src/org/catrobat/catroid/uitest/content/brick/ChangeColorByNBrickTest.java]: One could argue why this behaviour is there in the first place, but I have to agree: at least its clear why things are as they are...

File: catroid/src/org/catrobat/catroid/content/actions/ChangeColorByNAction.java (hunk 3 above)
Comment (84n4n4) [inline catroid/src/org/catrobat/catroid/content/actions/ChangeColorByNAction.java]: you can use the same logging construct as in all other classes, defining:
private static final String TAG = ChangeColorByNAction.class.getSimpleName();
and then just using
Log.d(TAG,"some log message")

**Comment by thmq** [comment]
please look at the bot's suggestions regarding the code format and the PMD warning [here](https://jenkins.catrob.at/job/CatroidCheck/103/console).

**Pull request closed without merging**